drv/lpc55-sprot-server/src/handler.rs

```diff
@@ -116,217 +121,194 @@ impl Handler {
+            }
+            MsgType::UpdFinishImageUpdateReq => {
+                let rsp: UpdateRspHeader = self
+                    .update
+                    .finish_image_update()
+                    .map(|_| None)
+                    .map_err(|e| e.into());
+                hubpack::serialize(tx_payload, &rsp)?
+            }
+            MsgType::UpdCurrentVersionReq => {
+                let rsp: Result<ImageVersion, u32> =
+                    self.update.current_version().map_err(|e| e.into());
+                hubpack::serialize(tx_payload, &rsp)?
+            }
+            MsgType::SinkReq => {
+                // The first two bytes of a SinkReq payload are the U16
+                // mod 2^16 sequence number.
+                tx_payload[0..2].copy_from_slice(&rx_payload[0..2]);
+                2
+            }
+            // All of the unexpected messages
+            MsgType::Invalid
+            | MsgType::EchoRsp
+            | MsgType::ErrorRsp
+            | MsgType::SinkRsp
+            | MsgType::SprocketsRsp
+            | MsgType::StatusRsp
+            | MsgType::UpdBlockSizeRsp
+            | MsgType::UpdPrepImageUpdateRsp
+            | MsgType::UpdWriteOneBlockRsp
+            | MsgType::UpdAbortUpdateRsp
+            | MsgType::UpdFinishImageUpdateRsp
+            | MsgType::UpdCurrentVersionRsp
+            | MsgType::Unknown => {
+                status.rx_invalid = status.rx_invalid.wrapping_add(1);
+                return Err(SprotError::BadMessageType);
+            }
+        };
+
+        Ok((req_msgtype_to_rsp_msgtype(rxmsg.0.msgtype), size))
+    }
+}
+
+// Translate a request msg type to a response msg type
+fn req_msgtype_to_rsp_msgtype(msgtype: MsgType) -> MsgType {
+    match msgtype {
+        MsgType::EchoReq => MsgType::EchoRsp,
+        MsgType::StatusReq => MsgType::StatusRsp,
+        MsgType::SprocketsReq => MsgType::SprocketsRsp,
+        MsgType::UpdBlockSizeReq => MsgType::UpdBlockSizeRsp,
+        MsgType::UpdPrepImageUpdateReq => MsgType::UpdPrepImageUpdateRsp,
+        MsgType::UpdWriteOneBlockReq => MsgType::UpdWriteOneBlockRsp,
+        MsgType::UpdAbortUpdateReq => MsgType::UpdAbortUpdateRsp,
+        MsgType::UpdFinishImageUpdateReq => MsgType::UpdFinishImageUpdateRsp,
+        MsgType::UpdCurrentVersionReq => MsgType::UpdCurrentVersionRsp,
+        MsgType::SinkReq => MsgType::SinkRsp,
+
+        // All of the unexpected messages
+        MsgType::Invalid
+        | MsgType::EchoRsp
+        | MsgType::ErrorRsp
+        | MsgType::SinkRsp
+        | MsgType::SprocketsRsp
+        | MsgType::StatusRsp
+        | MsgType::UpdBlockSizeRsp
+        | MsgType::UpdPrepImageUpdateRsp
+        | MsgType::UpdWriteOneBlockRsp
+        | MsgType::UpdAbortUpdateRsp
+        | MsgType::UpdFinishImageUpdateRsp
+        | MsgType::UpdCurrentVersionRsp
+        | MsgType::Unknown => {
+            panic!("MsgType is not a request: {}", msgtype as u8)
```

Comment:
Needing to duplicate all of these here and on lines 260-273 above is pretty unfortunate. Could we "inline" this function by having the match in `self.run()` give back a tuple of response type and length?